project1/server/agent.cpp

```diff
@@ -4,231 +4,415 @@
+                 std::shared_ptr<server_tasks::CommandIDs> active_commands)
+            : client_fd_(client_fd), active_commands_(active_commands), file_handler_(std::move(file_handler)) {}
+
+    Agent::Agent(int client_fd,
+                 std::shared_ptr<server_tasks::CommandIDs> active_commands) :
+            client_fd_(client_fd), active_commands_(active_commands) {}
+
+    Agent::~Agent() {
+        // Close the socket.
+        close(client_fd_);
+    }
+
+    bool Agent::Handle() {
+        ClientState client_state = ClientState::ACTIVE;
+
+
+        while (client_state == ClientState::ACTIVE) {
+            // Get the next message from the socket.
+            Request message;
+            client_state = ReadNextMessage(&message);
+
+            if (client_state == ClientState::ACTIVE) {
+                // Handle the message.
+                client_state = DispatchMessage(message);
+            }
+        }
+
+        return client_state != ClientState::ERROR;
+    }
+
+    Agent::ClientState Agent::ReadFileContents(ftp_messages::FileContents *fc, uint16_t command_id) {
+
+        while (!fc_parser_.HasCompleteMessage()) {
+            incoming_message_buffer_.resize(kClientBufferSize);
+
+            uint8_t buf[1000];
+
+            // Read 1000 bytes from the socket
+            const auto bytes_read =
+                    recv(client_fd_, buf, 1000, 0);
+
+            if (bytes_read < 0) {
+                // Failed to read anything.
+                perror("Failed to read from client socket");
+                fc_parser_.ResetParser();
+                return ClientState::ERROR;
+            } else if (bytes_read == 0) {
+                // Client has disconnected nicely.
+                std::cout << "Detected client disconnect." << std::endl;
+                return ClientState::DISCONNECTED;
+            }
+
+            bool terminated = !active_commands_->Contains(command_id);
+
+            // only continue if this put request is an active command
+            if (!terminated) {
+
+                // place data into the incoming message buffer
+                for (int i = 0; i < 1000; i++) {
+                    incoming_message_buffer_.push_back(buf[i]);
+                }
+
+            } else {
+                std::cout << "Command #" << command_id << " successfully terminated." << std::endl;
+            }
+
+            // The parser assumes that the entire vector contains valid data, so limit
+            // the size.
+            incoming_message_buffer_.resize(bytes_read);
+
+            // Add the data to the parser.
+            fc_parser_.AddNewData(incoming_message_buffer_);
+        }
+        // Get the parsed message.
+        if (!fc_parser_.GetMessage(fc)) {
+            std::cerr << "Failed to get the parsed message." << std::endl;
+            return ClientState::ERROR;
+        }
+        return ClientState::ACTIVE;
+    }
+
+    Agent::ClientState Agent::ReadNextMessage(Request *message) {
```

Comment:
This method seems redundant, (and will only get even more so if you take my suggestions above). Would it be possible to simply combine them into one and make the termination check optional? Or, you can break out the code in the loop into a new function, and call it from both places.

Reply:
It does seem redundant, but I think it's necessary to have two seperate, because, during `PUT` requests, we have to first recieve a `PutRequest` , send a `PutResponse`, and then explicitly wait for a `FileContents Message`.  These methods also use two different `MessageParser` objects. For these reasons I'm currently not seeing an intuitive way to combine these functions. 

As far as breaking out of the loop into a new funciton, I'd only be able to encapsulate a small amount of code into it, so that's not even worth it.